### ISSUE TITLE
[Research] Generate Oculus Store compatible APKs

### ISSUE THREAD

**Issue by FluorescentHallucinogen** [GoogleChromeLabs/bubblewrap#615]
**Initial request**

https://twitter.com/ThePra9/status/1455899334365618181

**Current plan**

1. Check if APK generated by Oculus CLI uses TWA (Trusted Web Activity):
1.1. Generate APK using Oculus CLI:
```sh
./ovr-platform-util create-pwa
  -o output.apk
  --android-sdk /opt/android_sdk
  https://example.com/manifest.webmanifest
```
Or to use local manifest file:
```sh
--manifest-content-file manifest.webmanifest
  https://example.com/
```
2. Check if Oculus supports TWA (Trusted Web Activity):
2.1. Generate APK using Bubblewrap (PWABuilder), try to publish it to Oculus Store via Oculus CLI or web interface (see https://developer.oculus.com/resources/publish-reference-platform-command-line-utility/#upload-quest) and make some changes to APK source code to fix errors/warnings (see https://developer.oculus.com/resources/publish-mobile-manifest/#specs).
2.2. Compare APK generated by Oculus CLI with APK generated by Bubblewrap (`AndroidManifest.xml`, etc.).
2.3. Check APK with fixes published to Oculus Store on real Oculus device.

3. Create a pull request for Bubblewrap.

4. Create a pull request for PWABuilder.

5. Write docs. E.g. document new vendor-specific proprietary web app manifest members such as:

```json
"ovr_scope_extensions": [
  {"pattern": "*.example.com"}
]
```

and

```json
"ovr_multi_tab_enabled" : true
```

**Notes**

This should be an option to not affect APKs for other app stores.

**Comment by andreban** [comment]
On 1, if someone can share an APK generated by the Oculus CLI I'm happy to check it out.

**Comment by FluorescentHallucinogen** [comment]
Hmm, for some reason I have no `create-pwa` command in `ovr-platform-util`: https://twitter.com/alexey_rodionov/status/1456689833380827137. 🤔

@jacobrossi PTAL.

**Comment by arpu** [comment]
the cli is https://developer.oculus.com/documentation/web/pwa-packaging/#download-the-cli
mac and widows version the zip has a pwa-template.apk included

**Comment by FluorescentHallucinogen** [comment]
@andreban

Here's APK: https://drive.google.com/file/d/1IAw8M0kc695GiawnBAvijfqzohrWHPbJ

I'm almost 100% sure it's not a TWA. :)

**Comment by andreban** [comment]
Interesting - despite being ~1mb in side, the Android Manifest doesn't have any Activity and neither does the code in `classes.dex`. They could make it much smaller by dropping the androidx dependencies.

Overall, their approach seems to be similar to how Trusted Web Activity works on Chrome OS on some scenarios - when the application is installed, Chrome OS reads the APK and extracts the information shipped within the app to generate an installed PWA - Bubblewrap [ships a copy of the Web Manifest inside the APK](https://github.com/GoogleChromeLabs/bubblewrap/blob/83435bf4d73ad74f4d758d1f1f858b51268b3ea7/packages/core/src/lib/TwaGenerator.ts#L263-L288).

**Comment by FluorescentHallucinogen** [comment]
> Interesting - despite being ~1mb in side, the Android Manifest doesn't have any Activity and neither does the code in `classes.dex`. They could make it much smaller by dropping the androidx dependencies.

@jacobrossi @cabanier FYI. :wink:

BTW, @jacobrossi @cabanier Does Oculus Browser supports TWA (Trusted Web Activity)?

**Comment by JohnRSim** [comment]
Anyone made any progress - any examples walkthroughs?

**Comment by FluorescentHallucinogen** [comment]
I'm still working on it. The problem is I don't have my own Oculus device. So I ask my friend who has the device to check some things for me.

Along the way, I check what PWA features on Oculus works and what doesn't. See https://github.com/immersive-web/webxr-samples/pull/135. :wink:

That's why it's so slow. :)

**Comment by JohnRSim** [comment]
If you need another device I have Quest2 I can test with for you.

**Comment by G8342** [comment]
>

**Comment by FluorescentHallucinogen** [comment]
Good news! I finally own Quest 2. 😊

ICYMI, preview version of PWABuilder (https://preview.pwabuilder.com) offers packaging PWA for Oculus Store. 😉

@tomayac cc.

It uses `ovr-platform-util` under the hood. Here's the repo: https://github.com/pwa-builder/pwabuilder-oculus.

**Comment by FluorescentHallucinogen** [comment]
I've also checked TWA on Quest 2. I've sideloaded APK via ADB and opened it via "Unknown Sources" section. And it just opened in Oculus Browser just like regular website (not in standalone mode).

I've also sideloaded Google Chrome APK to Quest 2. :) If Chrome is installed, then TWA open in it in standalone mode (instead of Oculus Browser).

**Comment by tomayac** [comment]
Oh, thanks for sharing, @FluorescentHallucinogen. That's really a nice feature addition to PWABuilder!

**Comment by FluorescentHallucinogen** [comment]
For anyone who's interested:

1.

Packaging PWAs for Quest with PWABuilder is officially released in stable channel: https://twitter.com/pwabuilder/status/1529191206919491586.

It's the recommended way to package PWAs for Quest: https://developer.oculus.com/documentation/web/pwa-building-with-pwabuilder/.

So I've updated @tomayac's “PWAs on Oculus Quest 2” blog post: https://web.dev/pwas-on-oculus-2/.

2.

I've reduced the size of `pwa-template.apk` by 1000x, from 1.2 MB to 1.2 kB! See my PWABuilder pull request for details: https://github.com/pwa-builder/pwabuilder-oculus/pull/7.

The `pwa-template.apk` is used in `ovr-platform-util` CLI that is used in PWABuilder under the hood for Packaging PWAs for Quest.

3.

I've finally got my hands on working on pull request that adds to Bubblewrap a flag for generating **universal** APKs compatible with Quest that open Trusted Web Activity (TWA) on regular Android devices and open Meta Quest Browser on Quest devices. This will allow to abandon `ovr-platform-util` in favor of Bubblewrap.
